src/contract.ts

```diff
@@ -1,376 +1,479 @@
+        const input = this.interface.encodeFunctionData(fragment, args);
+
+        const data = Buffer.concat([
+            this.storage.toBuffer(), //                     storage @FIXME: these comments are kind of useless
+            sender.toBuffer(), //                           sender  @FIXME: better to explain why, not what
+            Buffer.from(numToPaddedHex(value), 'hex'), //   value
+            Buffer.from('00000000', 'hex'), //              hash
+            encodeSeeds(seeds), //                          seeds
+            Buffer.from(input.replace('0x', ''), 'hex'), // input
+        ]);
+
+        const keys = [
+            // @FIXME: should all these PDAs really be writable?
+            ...programDerivedAddresses.map(({ address }) => ({
+                pubkey: address,
+                isSigner: false,
+                isWritable: true,
+            })),
+            {
+                pubkey: this.storage,
+                isSigner: false,
+                isWritable: true,
+            },
+            {
+                pubkey: SYSVAR_CLOCK_PUBKEY,
+                isSigner: false,
+                isWritable: false,
+            },
+            {
+                pubkey: PublicKey.default,
+                isSigner: false,
+                isWritable: false,
+            },
+            ...accounts.map((pubkey) => ({
+                pubkey,
+                isSigner: false,
+                isWritable: false,
+            })),
+            ...writableAccounts.map((pubkey) => ({
+                pubkey,
+                isSigner: false,
+                isWritable: true,
+            })),
+        ];
+
+        const transaction = new Transaction().add(
+            new TransactionInstruction({
+                keys,
+                programId: this.program,
+                data,
+            })
+        );
+
+        // If the function is read-only, simulate the transaction to get the result
+        const { logs, encoded, computeUnitsUsed } =
+            simulate || fragment.stateMutability === 'view' || fragment.stateMutability === 'pure'
+                ? await simulateTransactionWithLogs(this.connection, transaction, [payer, ...signers])
+                : await sendAndConfirmTransactionWithLogs(
+                      this.connection,
+                      transaction,
+                      [payer, ...signers],
+                      confirmOptions
+                  );
+
+        const events = this.parseLogsEvents(logs);
+
+        const length = fragment.outputs?.length;
+        let result: Result | null = null;
+
+        if (length) {
+            if (!encoded) throw new Error('MISSING_RETURN_DATA'); // @FIXME: add error types
+            result = this.interface.decodeFunctionResult(fragment, encoded);
+        }
+
+        if (returnResult) return result && length === 1 ? result[0] : result;
+        return { result, logs, events, computeUnitsUsed };
```

Comment:
We now simulate if the function is an accessor and return just the value unless the user calls `contract.functions.name(...)` to get the result metadata.